From daf9cbed5d4d15ed6354d40dcd9fef5fe1a33f73 Mon Sep 17 00:00:00 2001
From: Yi Xu <yiyix@chromium.org>
Date: Mon, 7 Mar 2022 22:23:46 +0000
Subject: [PATCH] Fix letterSpacing/wordSpacing for Canvas.Style

Canvas Style updates the fontDescription with the updated letterSpacing
and wordSpacing values, then SetFont resets the letterSpacing and
wordSpacing values to the ones in canvasRenderingContext2D.

The result should be the sum of the two.

Bug: 1296564

(cherry picked from commit 8096c12f2e6e53d4b2fa9811ddcda807da042708)

Change-Id: I6ec70e649f61d52b329e8252fe158a23566fb612
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3479474
Reviewed-by: Fernando Serboncini <fserb@chromium.org>
Reviewed-by: Justin Novosad <junov@chromium.org>
Commit-Queue: Yi Xu <yiyix@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#973927}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3507613
Reviewed-by: Aaron Krajeski <aaronhk@chromium.org>
Cr-Commit-Position: refs/branch-heads/4844@{#1003}
Cr-Branched-From: 007241ce2e6c8e5a7b306cc36c730cd07cd38825-refs/heads/main@{#961656}
---
 .../canvas_rendering_context_2d_state.cc      | 32 ++++++++++++++-----
 .../canvas_rendering_context_2d_state.h       |  2 ++
 2 files changed, 26 insertions(+), 8 deletions(-)

--- a/third_party/blink/renderer/modules/canvas/canvas2d/canvas_rendering_context_2d_state.cc
+++ b/third_party/blink/renderer/modules/canvas/canvas2d/canvas_rendering_context_2d_state.cc
@@ -83,6 +83,8 @@
       is_transform_invertible_(true),
       has_clip_(false),
       has_complex_clip_(false),
+      letter_spacing_is_set_(false),
+      word_spacing_is_set_(false),
       fill_style_dirty_(true),
       stroke_style_dirty_(true),
       line_dash_dirty_(false),
@@ -148,6 +150,8 @@
       is_transform_invertible_(other.is_transform_invertible_),
       has_clip_(other.has_clip_),
       has_complex_clip_(other.has_complex_clip_),
+      letter_spacing_is_set_(other.letter_spacing_is_set_),
+      word_spacing_is_set_(other.word_spacing_is_set_),
       fill_style_dirty_(other.fill_style_dirty_),
       stroke_style_dirty_(other.stroke_style_dirty_),
       line_dash_dirty_(other.line_dash_dirty_),
@@ -313,15 +317,23 @@
       1.0f /*Deliberately ignore zoom on the canvas element*/);
   conversion_data.SetFontSizes(font_size);
 
-  // Convert word spacing to pixel length and set it in font_description.
-  float word_spacing_in_pixel =
-      conversion_data.ZoomedComputedPixels(word_spacing_, word_spacing_unit_);
-  font_description.SetWordSpacing(word_spacing_in_pixel);
-
-  // Convert letter spacing to pixel length and set it in font_description.
-  float letter_spacing_in_pixel = conversion_data.ZoomedComputedPixels(
-      letter_spacing_, letter_spacing_unit_);
-  font_description.SetLetterSpacing(letter_spacing_in_pixel);
+  // If wordSpacing is set in CanvasRenderingContext2D, then update the
+  // information in fontDescription.
+  if (word_spacing_is_set_) {
+    // Convert word spacing to pixel length and set it in font_description.
+    float word_spacing_in_pixel =
+        conversion_data.ZoomedComputedPixels(word_spacing_, word_spacing_unit_);
+    font_description.SetWordSpacing(word_spacing_in_pixel);
+  }
+
+  // If wordSpacing is set in CanvasRenderingContext2D, then update the
+  // information in fontDescription.
+  if (letter_spacing_is_set_) {
+    // Convert letter spacing to pixel length and set it in font_description.
+    float letter_spacing_in_pixel = conversion_data.ZoomedComputedPixels(
+        letter_spacing_, letter_spacing_unit_);
+    font_description.SetLetterSpacing(letter_spacing_in_pixel);
+  }
 
   font_ = Font(font_description, selector);
   realized_font_ = true;
@@ -773,6 +785,8 @@
 void CanvasRenderingContext2DState::SetLetterSpacing(
     const String& letter_spacing) {
   DCHECK(realized_font_);
+  if (!letter_spacing_is_set_)
+    letter_spacing_is_set_ = true;
   if (parsed_letter_spacing_ == letter_spacing)
     return;
   float num_spacing;
@@ -794,6 +808,8 @@
 
 void CanvasRenderingContext2DState::SetWordSpacing(const String& word_spacing) {
   DCHECK(realized_font_);
+  if (!word_spacing_is_set_)
+    word_spacing_is_set_ = true;
   if (parsed_word_spacing_ == word_spacing)
     return;
   float num_spacing;
--- a/third_party/blink/renderer/modules/canvas/canvas2d/canvas_rendering_context_2d_state.h
+++ b/third_party/blink/renderer/modules/canvas/canvas2d/canvas_rendering_context_2d_state.h
@@ -322,6 +322,8 @@
   bool is_transform_invertible_ : 1;
   bool has_clip_ : 1;
   bool has_complex_clip_ : 1;
+  bool letter_spacing_is_set_ : 1;
+  bool word_spacing_is_set_ : 1;
   mutable bool fill_style_dirty_ : 1;
   mutable bool stroke_style_dirty_ : 1;
   mutable bool line_dash_dirty_ : 1;
